primitive attribute assignment

avikchaudhuri opened the pull request:
This PR ensures that assignment of attributes of primitive type work without needing any code changes in non-strict mode. (In a previous PR we banned attribute assignments of tensor type unless such attributes are registered as buffers.)

While strict mode errors on (all) attribute assignments, non-strict doesn't care, so one might assume that this kind of attribute assignment should already work in non-strict. However, there's a problem: we run through the program once for metadata collection and then run through it again for tracing, so the values observed during tracing (and potentially burned into the graph) do not reflect what should have been observed had the metadata collection pass not run. 

So the only thing this PR needs to do is restore values of assigned attributes of primitive type once the metadata collection pass has run. We do this by moving the attribute assignment detecting context manager from the overall `aot_export` call in `_trace.py` to the metadata collection pass in `aot_autograd.py`, and extending it. The rest of the PR moves some utils around.
 
Differential Revision: D56047952

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -526,6 +526,24 @@ def forward(self, x):
+            def forward(self, x):
+                self.x = self.x + 3
+                self.y["x"] = self.y["x"] + 4
+                return x + self.x + self.y["x"]
```
This should now generate code of the form `x + 4 + 6` (where 4 is inlined `self.x` and 6 is inlined `self.y["x"]`) but without this PR the generated code would be of the form `x + 7 + 10`.

avikchaudhuri commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1254,5 +1261,79 @@ def _aot_export_function(
+                    assigned_tensor_attributes.append(
+                        f"self.{attr.key}{pytree.keystr(rest)}"
+                    )
+                elif isinstance(v, (int, float, str, bool)):
```
main change is here

angelayi commented on test/export/test_export.py:
```diff
@@ -526,6 +526,24 @@ def forward(self, x):
+                self.y["x"] = self.y["x"] + 4
+                return x + self.x + self.y["x"]
+
+        ep = torch.export.export(M(), (torch.randn(2, 3),), strict=False)
```
Do you want to add an assertion that after this `M().x` and `M().y` stays the same?

avikchaudhuri replied:
They doesn't stay the same. I'm not restoring anything after export, only after the metadata collection pass. Whether to restore the state of the original module or not is an orthogonal design decision. 

(This doesn't affect the exported program, of course, because the observed state is burned in.)

angelayi replied:
ooh ok

angelayi commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1254,5 +1261,79 @@ def _aot_export_function(
+        elif isinstance(k, pytree.MappingKey):
+            obj[k.key] = v
+        elif isinstance(k, pytree.SequenceKey):
+            obj[k.idx] = v
```
Does this work for tuples?

avikchaudhuri replied:
Nice, it breaks with tuples. :) I'll do the restore differently.

angelayi reviewed (approved):
very cool!

avikchaudhuri committed:
primitive attribute assignment (#123898)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/123898

This PR ensures that assignment of attributes of primitive type work without needing any code changes in non-strict mode. (In a previous PR we banned attribute assignments of tensor type unless such attributes are registered as buffers.)

While strict mode errors on (all) attribute assignments, non-strict doesn't care, so one might assume that this kind of attribute assignment should already work in non-strict. However, there's a problem: we run through the program once for metadata collection and then run through it again for tracing, so the values observed during tracing (and potentially burned into the graph) do not reflect what should have been observed had the metadata collection pass not run.

So the only thing this PR needs to do is restore values of assigned attributes of primitive type once the metadata collection pass has run. We do this by moving the attribute assignment detecting context manager from the overall aot_export call in _trace.py to the metadata collection pass in aot_autograd.py, and extending it. The rest of the PR moves some utils around.

Test Plan: added test

Differential Revision: D56047952

avikchaudhuri commented:
@pytorchbot merge
